parser/input.go

```diff
@@ -0,0 +1,383 @@
+	if offset == 0 {
+		return 0
+	}
+	if offset < 0 {
+		offset++
+	}
+	pos := s.pos + offset - 1
+	if pos < 0 || pos >= len(s.arr) {
+		return antlr.TokenEOF
+	}
+	return int(uint(s.arr[pos]))
+}
+
+func (s *supplementalCodePointBuffer) LT(offset int) int {
+	return s.LA(offset)
+}
+
+func (s *supplementalCodePointBuffer) Mark() int {
+	return -1
+}
+
+func (s *supplementalCodePointBuffer) Release(marker int) {}
+
+func (s *supplementalCodePointBuffer) Index() int {
+	return s.pos
+}
+
+func (s *supplementalCodePointBuffer) Seek(index int) {
+	if index <= s.pos {
+		s.pos = index
+		return
+	}
+	if index < len(s.arr) {
+		s.pos = index
+	} else {
+		s.pos = len(s.arr)
+	}
+}
+
+func (s *supplementalCodePointBuffer) Size() int {
+	return len(s.arr)
+}
+
+func (s *supplementalCodePointBuffer) GetSourceName() string {
+	return s.src
+}
+
+func (s *supplementalCodePointBuffer) GetText(start, stop int) string {
+	if stop >= len(s.arr) {
+		stop = len(s.arr) - 1
+	}
+	if start >= len(s.arr) {
+		return ""
+	}
+	return string(s.arr[start : stop+1])
+}
+
+func (s *supplementalCodePointBuffer) GetTextFromTokens(start, stop antlr.Token) string {
+	if start != nil && stop != nil {
+		return s.GetText(start.GetTokenIndex(), stop.GetTokenIndex())
+	}
+	return ""
+}
+
+func (s *supplementalCodePointBuffer) GetTextFromInterval(i *antlr.Interval) string {
+	return s.GetText(i.Start, i.Stop)
+}
+
+func (s *supplementalCodePointBuffer) String() string {
+	return string(s.arr)
+}
+
+var _ antlr.CharStream = &supplementalCodePointBuffer{}
+
+// newCodePointBuffer returns an efficient implementation of antlr.CharStream for the given text
+// based on the ranges of the encoded code points contained within.
+//
+// Code points are represented as an array of byte, uint16, or rune. This approach ensures that
+// each index represents a code point by itself without needing to use an array of rune. At first
+// we assume all code points are less than or equal to '\u007f'. If this holds true, the
+// underlying storage is a byte array containing only ASCII characters. If we encountered a code
+// point above this range but less than or equal to '\uffff' we allocate a uint16 array, copy the
+// elements of previous byte array to the uint16 array, and continue. If this holds true, the
+// underlying storage is a uint16 array containing only Unicode characters in the Basic Multilingual
+// Plane. If we encounter a code point above '\uffff' we allocate an rune array, copy the previous
+// elements of the byte or uint16 array, and continue. The underlying storage is an rune array
+// containing any Unicode character.
+//
+// TODO: upstream this to ANTLRv4 Go runtime?
+func newCodePointBuffer(text, desc string) antlr.CharStream {
+	idx := 0
+	buf8 := make([]byte, 0, len(text))
+	for idx < len(text) {
+		r, s := utf8.DecodeRuneInString(text[idx:])
+		idx += s
+		if r < utf8.RuneSelf {
+			buf8 = append(buf8, byte(r))
+			continue
+		}
+		if r <= 0xffff {
```

Comment:
Rather than replicating the for-loops, consider using labels to jump to the appropriate loop.

```go
Copy16:
  buf16 = append(buf16, uint16(r))
  for idx < len(text) {
     r, s = utf8.DecodeRuneInString(text[idx:])
     idx += s				
     if r > 0xffff {
        buf32 := make([]rune, len(buf16), len(text))
        for i, v := range buf16 {
            buf32[i] = rune(uint32(v))
        }
        goto Copy32
     }
     buf16 = append(buf16, uint16(r))
     ...
  }
  return ...
Copy32:
  buf32 = append(buf32, r)
  for idx < len(text) {
  ...
  }
  return ...
```

Reply:
Done.